worker/modelcache/worker.go

```diff
@@ -0,0 +1,246 @@
+	}
+	if c.StatePool == nil {
+		return errors.NotValidf("missing state pool")
+	}
+	if c.PrometheusRegisterer == nil {
+		return errors.NotValidf("missing prometheus registerer")
+	}
+	if c.Cleanup == nil {
+		return errors.NotValidf("missing cleanup func")
+	}
+	return nil
+}
+
+type cacheWorker struct {
+	config     Config
+	catacomb   catacomb.Catacomb
+	controller *cache.Controller
+	changes    chan interface{}
+	watcher    *state.Multiwatcher
+	mu         sync.Mutex
+}
+
+// NewWorker creates a new cacheWorker, and starts an
+// all model watcher.
+func NewWorker(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+	w := &cacheWorker{
+		config:  config,
+		changes: make(chan interface{}),
+	}
+	controller, err := cache.NewController(
+		cache.ControllerConfig{
+			Changes: w.changes,
+			Notify:  config.Notify,
+		})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	w.controller = controller
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.controller},
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return w, nil
+}
+
+// Report returns information that is used in the dependency engine report.
+func (c *cacheWorker) Report() map[string]interface{} {
+	return c.controller.Report()
+}
+
+func (c *cacheWorker) loop() error {
+	defer c.config.Cleanup()
+	pool := c.config.StatePool
+
+	allWatcherStarts := prometheus.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "juju_worker_modelcache",
+			Name:      "watcher_starts",
+			Help:      "The number of times the all model watcher has been started.",
+		},
+	)
+
+	collector := cache.NewMetricsCollector(c.controller)
+	c.config.PrometheusRegisterer.Register(collector)
+	c.config.PrometheusRegisterer.Register(allWatcherStarts)
+	defer c.config.PrometheusRegisterer.Unregister(allWatcherStarts)
+	defer c.config.PrometheusRegisterer.Unregister(collector)
+
+	watcherChanges := make(chan []multiwatcher.Delta)
+	// This worker needs to be robust with respect to the multiwatcher
+	// errors. If we get an unexpected error we should get a new allWatcher.
+	// We don't want a weird error in the multiwatcher taking down the apiserver,
+	// which is what would happen if this worker errors out.
+	// We do need to consider cache invalidation for multiwatcher entities
+	// that may be in our cache but when we restart the watcher, they aren't there.
+	// Cache invalidtion is a hard problem, but here at least we should perhaps
+	// be able to do some form of mark and sweep. When we create a new watcher
+	// we should mark entities in the controller, and when we are done with the
+	// first call to Next(), which returns the state of the world, we can issue
+	// a sweep to remove anything that wasn't updated since the Mark.
+	// TODO: This is left for upcoming work.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer func() {
+		c.mu.Lock()
+		c.watcher.Stop()
```

Comment:
Why do you need `c.watcher.Stop()` here? Under what circumstances won't the goroutine below stop it?

Reply:
When the worker is being stopped from the outside, the dying channel is closed. This causes the watcher to be stopped, which triggers an error in the Next call, which causes the handling goroutine to terminate.